builder/pipeline.cc

```diff
@@ -258,118 +263,164 @@ bool operator==(const loop_id& a, const loop_id& b) {
+                                    std::map<const func*, loop_id>& compute_at_levels) {
+    // A tree which stores loop nest.
+    std::vector<loop_tree_node> loop_tree;
+    // Mapping between function and it's most innermost location.
+    std::map<const func*, int> func_to_loop_tree;
+    // Push the root loop.
+    loop_tree.push_back({-1, loop_id()});
+    
+    // Iterate over functions in topological order starting from the output and build a loop nest tree.
+    for (const auto& f: order) {
+      const auto& p = deps.find(f);
+      if (p != deps.end()) {
+        assert(!p->second.empty());
+        int parent_id = -1;
+
+        // If we have an explicitly set compute_at location we should use that.
+        if (f->compute_at()) {
+          // TODO(vksnk): check if compute_at is a valid location based on computed
+          // innermost location.
+          for (int ix = 0; ix < (int)loop_tree.size(); ix++) {
+            if (loop_tree[ix].loop == *f->compute_at()) {
+              parent_id = ix;
+            }
+          }
+          compute_at_levels[f] = *f->compute_at();
+        } else {
+          // Check all of the consumers and find their innermost locations.
+          std::vector<int> parent_ids;
+          for (const auto& d: p->second) {
+            const auto& node = func_to_loop_tree.find(d);
+            assert(node != func_to_loop_tree.end());
+            parent_ids.push_back(node->second);
+          }
+
+          if (parent_ids.size() == 1) {
+            // We have just one consumer, so use its innermost location.
+            parent_id = parent_ids[0];
+          } else {
+            // There are multiple consumers, so we need to find the least common ancestor
+            // of their innermost locations.
+            parent_id = lca(loop_tree, parent_ids);
+          }
 
-      if (i->producer()->compute_at()) {
-        if (!(*i->producer()->compute_at() == at)) {
-          // This shouldn't be computed here.
-          continue;
+          compute_at_levels[f] = loop_tree[parent_id].loop;
         }
-      } else if (!at.root()) {
-        // By default, we want to compute everything as soon as it is consumed, so it ends up in the innermost loop
-        // possible. But if we do this blindly, then unrelated pipelines that aren't consumed inside these loops at
-        // all will get generated here.
-        if (!consumed.count(i)) {
-          continue;
+
+        assert(parent_id != -1);
+
+        // Add loops for this function to the loop nest. The loops are defined
+        // from innermost to outermost, so iterate in reverse order.
+        for (int i = f->loops().size() - 1; i >= 0; i--) {
+          const auto& l = f->loops()[i];
+          loop_tree.push_back({parent_id, {f, l.var}});
+          parent_id = loop_tree.size() - 1;
+        }
+        func_to_loop_tree[f] = parent_id;
+      } else {
+        // This is an output so should be computed at root.
+        int parent_id = 0;
```

Comment:
This looks almost identical to the body of the then case above. Should `int parent_id = 0` be outside the if, change it in the then case, and then share this logic across both branches?

Reply:
Good point, refactored.